src/helpers.c

```diff
@@ -191,88 +191,80 @@ int system_argv(char *const argv[])
 	return status;
 }
 
-int system_argv_fd(char *const argv[], int newstdin, int newstdout, int newstderr)
+int system_argv_pipe(char *const lhscmd[], char *const rhscmd[])
 {
-	int child_exit_status;
-	pid_t pid;
-	int status = -1;
-
-	pid = fork();
-
-	if (pid == 0) { /* child */
-		if (newstdin >= 0) {
-			if (dup2(newstdin, STDIN_FILENO) == -1) {
-				LOG(NULL, "Could not redirect stdin", "");
-				assert(0);
-			}
-			close(newstdin);
+	pid_t childpid = fork();
+	if (childpid == -1) {
+		LOG(NULL, "Failed to create child process", "");
+		return -1;
+	} else if (childpid == 0) {
```

Comment:
Say there's a lot of data that needs to move between the lhs tar and the rhs tar.  Does waiting in the pipe monitor function prevent the rhs tar from being called in this case because the wait_process_terminate function does not acknowledge if a process was stopped or continued?  Because the lhs tar would have blocked once the pipe buffer would have filled up, waiting for the rhs tar to consume data from the pipe, right?  But the rhs tar would never be exec'd because we'd be waiting on the lhs tar?

Reply:
The 3 processes are able to run independently until they interact. The interactions are the pipe between lhs and rhs, where the lhs will pause until the rhs has read the data, and between the pipe monitor and the two processes it creates when it calls wait.

**The wait(2) call says 'put the pipemonitor to sleep until the lhs (or rhs, depending on what you are waiting for) has a change of state (essentially finished or get a signal)'.**

So the amount of data being transferred between the lhs and rhs might make a difference to the amount of clock time that passes before the pipe monitor is awoken, but nothing else.

The wait_process_terminate is ignores some state changes, although it could ignore more.

Reply:
Ah, ok.  I'm compiling and testing changes which incorporate all of your feedback and bring back the interface for managing file descriptors with commands.  If things go well, I'll add another commit which unifies all system_argv commands under common code and interfaces using small functions.